src/Form/ConstraintForm.php

```diff
@@ -0,0 +1,286 @@
+   *   The form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The from's current state.
+   *
+   * @return array
+   *   Renderable array for the settings wrapper.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   */
+  public function updateForm(array $form, FormStateInterface $form_state): array {
+    // Get type if exist.
+    if (($type = $this->entity->getType()) === '') {
+      $type = $form_state->getValue('type', '');
+    }
+
+    if ($type !== '') {
+      // Get plugin instance for the selected type.
+      $plugin_instance = $this->passwordConstraintPluginManager->createInstance($type, $this->entity->getConfiguration());
+
+      // Update description.
+      if ($this->entity->isNew()) {
+        // Unset description from the form state if the type was changed.
+        $input = &$form_state->getUserInput();
+        if (array_key_exists('descriptionSingular', $input)) {
+          unset($input['descriptionSingular']);
+        }
+
+        if (array_key_exists('descriptionPlural', $input)) {
+          unset($input['descriptionPlural']);
+        }
+
+        $form['wrapper']['descriptionSingular']['#default_value'] = $plugin_instance->defaultDescriptionSingular();
+        $form['wrapper']['descriptionPlural']['#default_value'] = $plugin_instance->defaultDescriptionPlural();
+      }
+      else {
+        $form['wrapper']['descriptionSingular']['#default_value'] = $this->entity->getDescriptionSingular();
+        $form['wrapper']['descriptionPlural']['#default_value'] = $this->entity->getDescriptionPlural();
+      }
+
+      // Get plugin specific form elements if type is set.
+      $form['wrapper']['settings'] += $plugin_instance->settingsForm();
+    }
+    else {
+      // Show an empty message if there is no type selected yet.
+      $form['wrapper']['settings']['empty'] = [
+        '#type' => 'item',
+        '#markup' => $this->t('Please select a constraint.'),
+      ];
+    }
+
+    return $form['wrapper'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state): void {
+    $type = $form_state->getValue('type');
+
+    if ($this->entity->isNew()) {
+      // If the constraint marked as unique, make sure that only one can be
+      // created for a given policy.
+      $plugin_definition = $this->passwordConstraintPluginManager->getDefinition($type);
+      if (!empty($plugin_definition['unique'])) {
+        $entities = $this->constraintConfigEntityStorage->loadByProperties([
+          'type' => $type,
+        ]);
+        if (array_key_exists($this->getRequest()->get('password_enhancements_policy')->id() . '.' . $type, $entities)) {
+          $form_state->setError($form['type'], $this->t('Only one instance can be created from the selected constraint.'));
+        }
+      }
+    }
+
+    // Validate plugin specific form elements.
+    if (!array_key_exists('empty', $form['wrapper']['settings'])) {
```

Comment:
Well, relying on this 
```
      $form['wrapper']['settings']['empty'] = [
        '#type' => 'item',
        '#markup' => $this->t('Please select a constraint.'),
      ];
```
form element definition is an interesting hacky solution. Could not you find a better way to identify in the submit/validate callbacks whether there is a selected constraint or not? Or could not you disable the save button until there is nothing selected?

Reply:
This is required because of the initial state of the form, basically it is just a notification for to see the settings the user has to select some type, the form itself is blocked until all of the required fields are not filled in.

Reply:
#2 can possibly improve this